src/backend/opencl/kernel/meanshift.cl

```diff
@@ -7,149 +7,115 @@
-
-    // gfor batch offsets
     unsigned b2 = get_group_id(0) / nBBS0;
     unsigned b3 = get_group_id(1) / nBBS1;
     __global const T* iptr = d_src + (b2 * iInfo.strides[2] + b3 * iInfo.strides[3] + iInfo.offset);
     __global T*       optr = d_dst + (b2 * oInfo.strides[2] + b3 * oInfo.strides[3]);
 
-    const int lx = get_local_id(0);
-    const int ly = get_local_id(1);
-
-    const int gx = get_local_size(0) * (get_group_id(0)-b2*nBBS0) + lx;
-    const int gy = get_local_size(1) * (get_group_id(1)-b3*nBBS1) + ly;
-
-    int s0 = iInfo.strides[0];
-    int s1 = iInfo.strides[1];
-    int d0 = iInfo.dims[0];
-    int d1 = iInfo.dims[1];
-    // pull image to local memory
-    for (int b=ly, gy2=gy; b<shrdLen; b+=get_local_size(1), gy2+=get_local_size(1)) {
-        // move row_set get_local_size(1) along coloumns
-        for (int a=lx, gx2=gx; a<shrdLen; a+=get_local_size(0), gx2+=get_local_size(0)) {
-            load2LocalMem(localMem, iptr, a, b, shrdLen, schStride, channels,
-                    d0, d1, gx2-radius, gy2-radius, ichStride, s1, s0);
-        }
-    }
-
-    int i   = lx + radius;
-    int j   = ly + radius;
-
-    barrier(CLK_LOCAL_MEM_FENCE);
+    const int gx = get_local_size(0) * (get_group_id(0)-b2*nBBS0) + get_local_id(0);
+    const int gy = get_local_size(1) * (get_group_id(1)-b3*nBBS1) + get_local_id(1);
 
     if (gx<iInfo.dims[0] && gy<iInfo.dims[1])
     {
-        float means[MAX_CHANNELS];
-        float centers[MAX_CHANNELS];
-        float tmpclrs[MAX_CHANNELS];
+        int i  = gx;
+        int j  = gy;
+
+        T centers[MAX_CHANNELS];
+        T tmpclrs[MAX_CHANNELS];
+
+        AccType means[MAX_CHANNELS];
 
-        // clear means and centers for this pixel
 #pragma unroll
-        for(int ch=0; ch<channels; ++ch) {
-            means[ch] = 0.0f;
-            centers[ch] = localMem[lIdx(i, j, shrdLen, 1)+ch*schStride];
-        }
+        for(int ch=0; ch<MAX_CHANNELS; ++ch)
+            centers[ch] = iptr[ i*iInfo.strides[0]+j*iInfo.strides[1]+ch*iInfo.strides[2] ];
+
+        const int dim0LenLmt = iInfo.dims[0]-1;
+        const int dim1LenLmt = iInfo.dims[1]-1;
 
         // scope of meanshift iterationd begin
         for(uint it=0; it<iter; ++it) {
 
-            int count   = 0;
+            int ocj = j;
+            int oci = i;
+            unsigned count   = 0;
+
             int shift_x = 0;
             int shift_y = 0;
 
+            for (int ch=0; ch<MAX_CHANNELS; ++ch)
+                means[ch] = 0;
+
             for(int wj=-radius; wj<=radius; ++wj) {
                 int hit_count = 0;
+                int tj = j + wj;
+
+                if (tj<0 || tj>dim1LenLmt) break;
 
                 for(int wi=-radius; wi<=radius; ++wi) {
 
-                    int tj = j + wj;
                     int ti = i + wi;
 
-                    // proceed
-                    float norm = 0.0f;
+                    if (ti<0 || ti>dim0LenLmt) break;
+
+                    AccType norm = 0;
```

Comment:
0.0f

Reply:
`AccType` can be double as well, if the input array is double.